File: data/src/main/java/com/arturogutierrez/openticator/storage/realm/mapper/AccountRealmMapper.java
package com.arturogutierrez.openticator.storage.realm.mapper;

import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.domain.account.model.OTPType;
import com.arturogutierrez.openticator.domain.issuer.model.Issuer;
import com.arturogutierrez.openticator.mapper.Mapper;
import com.arturogutierrez.openticator.storage.realm.model.AccountRealm;
import javax.inject.Inject;

public class AccountRealmMapper extends Mapper<Account, AccountRealm> {

  @Inject
  public AccountRealmMapper() {

  }

  @Override
  public AccountRealm transform(Account account) {
    AccountRealm accountRealm = null;

    if (account != null) {
      accountRealm = new AccountRealm();
      copyToAccountRealm(accountRealm, account);
    }

    return accountRealm;
  }

  @Override
  public Account reverseTransform(AccountRealm accountRealm) {
    Account account = null;

    if (accountRealm != null) {
      Issuer issuer = transformIssuer(accountRealm.getIssuer());
      account = new Account(accountRealm.getAccountId(), accountRealm.getName(),
          transformAccountType(accountRealm.getType()), accountRealm.getSecret(), issuer,
          accountRealm.getOrder());
    }

    return account;
  }

  public void copyToAccountRealm(AccountRealm accountRealm, Account account) {
    if (accountRealm == null || account == null) {
      return;
    }

    accountRealm.setAccountId(account.getAccountId());
    accountRealm.setName(account.getName());
    accountRealm.setSecret(account.getSecret());
    accountRealm.setIssuer(account.getIssuer().getIdentifier());
    accountRealm.setOrder(account.getOrder());
    accountRealm.setType(transformAccountType(account.getType()));
  }

  private OTPType transformAccountType(String otpType) {
    if (otpType.equals(AccountRealm.HOTP_TYPE)) {
      return OTPType.HOTP;
    }

    return OTPType.TOTP;
  }

  private String transformAccountType(OTPType otpType) {
    if (otpType == OTPType.HOTP) {
      return AccountRealm.HOTP_TYPE;
    }

    return AccountRealm.TOTP_TYPE;
  }

  private Issuer transformIssuer(String issuer) {
    return Issuer.Companion.fromString(issuer);
  }
}

File: data/src/main/kotlin/com/arturogutierrez/openticator/storage/realm/mapper/AccountRealmMapper.kt
package com.arturogutierrez.openticator.storage.realm.mapper

import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.domain.account.model.OTPType
import com.arturogutierrez.openticator.domain.issuer.model.Issuer
import com.arturogutierrez.openticator.mapper.Mapper
import com.arturogutierrez.openticator.storage.realm.model.AccountRealm
import javax.inject.Inject

class AccountRealmMapper @Inject constructor() : Mapper<Account, AccountRealm> {

  override fun transform(account: Account): AccountRealm {
    val accountRealm = AccountRealm()
    copyToAccountRealm(accountRealm, account)
    return accountRealm
  }

  override fun reverseTransform(accountRealm: AccountRealm): Account {
    val issuer = transformIssuer(accountRealm.issuer)
    val account = Account(accountRealm.accountId, accountRealm.name,
        transformAccountType(accountRealm.type), accountRealm.secret, issuer,
        accountRealm.order)
    return account
  }

  fun copyToAccountRealm(accountRealm: AccountRealm, account: Account) {
    with(accountRealm) {
      if (accountId != account.accountId) {
        accountId = account.accountId
      }
      name = account.name
      secret = account.secret
      issuer = account.issuer.identifier
      order = account.order
      type = transformAccountType(account.type)
    }
  }

  private fun transformAccountType(otpType: String): OTPType {
    if (otpType == AccountRealm.HOTP_TYPE) {
      return OTPType.HOTP
    }

    return OTPType.TOTP
  }

  private fun transformAccountType(otpType: OTPType): String {
    if (otpType === OTPType.HOTP) {
      return AccountRealm.HOTP_TYPE
    }

    return AccountRealm.TOTP_TYPE
  }

  private fun transformIssuer(issuer: String): Issuer {
    return Issuer.fromString(issuer)
  }
}
